tests/actor/test_resampling.py

```diff
@@ -122,39 +122,37 @@ async def test_single_request(
     resampling_req_chan = Broadcast[ComponentMetricRequest]("resample-req")
     resampling_req_sender = resampling_req_chan.new_sender()
 
-    resampling_actor = ComponentMetricsResamplingActor(
+    async with ComponentMetricsResamplingActor(
         channel_registry=channel_registry,
         data_sourcing_request_sender=data_source_req_chan.new_sender(),
         resampling_request_receiver=resampling_req_chan.new_receiver(),
         config=ResamplerConfig(
             resampling_period=timedelta(seconds=0.2),
             max_data_age_in_periods=2,
         ),
-    )
-
-    subs_req = ComponentMetricRequest(
-        namespace="Resampling",
-        component_id=9,
-        metric_id=ComponentMetricId.SOC,
-        start_time=None,
-    )
-
-    await resampling_req_sender.send(subs_req)
-    data_source_req = await data_source_req_recv.receive()
-    assert data_source_req is not None
-    assert data_source_req == dataclasses.replace(
-        subs_req, namespace="Resampling:Source"
-    )
-
-    await _assert_resampling_works(
-        channel_registry,
-        fake_time,
-        resampling_chan_name=subs_req.get_channel_name(),
-        data_source_chan_name=data_source_req.get_channel_name(),
-    )
-
-    await resampling_actor._stop()  # type: ignore # pylint: disable=no-member,protected-access
-    await resampling_actor._resampler.stop()  # pylint: disable=protected-access
+    ) as resampling_actor:
+        subs_req = ComponentMetricRequest(
+            namespace="Resampling",
+            component_id=9,
+            metric_id=ComponentMetricId.SOC,
+            start_time=None,
+        )
+
+        await resampling_req_sender.send(subs_req)
+        data_source_req = await data_source_req_recv.receive()
+        assert data_source_req is not None
+        assert data_source_req == dataclasses.replace(
+            subs_req, namespace="Resampling:Source"
+        )
+
+        await _assert_resampling_works(
+            channel_registry,
+            fake_time,
+            resampling_chan_name=subs_req.get_channel_name(),
+            data_source_chan_name=data_source_req.get_channel_name(),
+        )
+
+        await resampling_actor._resampler.stop()  # pylint: disable=protected-access
```

Comment:
Not so simple: This is related to the comment in the resampling actor's `_run()` cleanup (`finally`):

```py
            # XXX: Here we should probably do a:  pylint: disable=fixme
            # await self._resampler.stop()
            # But since the actor will be restarted, the internal state would
            # be broken if we stop the resampler.
            #
            # We have an even bigger problem with this naive restarting
            # approach, as restarting this actor without really resetting its
            # state would be mostly the same as not really leaving the run()
            # method and just swallow any exception, which doesn't look super
            # smart.
```

This is why I'm planning to move the restart logic from the `Actor` to the `run()` method. Since the actor is monitoring himself, there is one task that is special, as we can't kill it because is the task monitoring the `_run()` method and restarting it. The self-auto-restarting makes everything more complicated and it is not very resilient,as the actor's internal state is not reset, so if it was left in an inconsistent state because of some bug, re-starting the `_run()` won't fix it. Actually restarting it in `run()` fixes part of the problem but not this, maybe we need some explicit `restart()` method that cleans up the internal state.

So I would leave this for future PRs, I also didn't converted more classes to background services (the resampler and its helper classes should be also background services) because they are not breaking changes, and I wanted to prioritize breaking changes for 1.0.